compiler/src/iree/compiler/Dialect/LinalgExt/IR/TilingInterfaceImpl.cpp

```diff
@@ -2163,4 +2167,327 @@ LogicalResult OnlineAttentionOp::getResultTilePosition(
+/// `numDims, numDims + 1, numDims + 2, ...., numDims + numSymbols - 1`.
+static SmallVector<AffineExpr> getDimExprsForSymbols(MLIRContext *context,
+                                                     unsigned numDims,
+                                                     unsigned numSymbols) {
+  return llvm::map_to_vector(
+      llvm::seq<unsigned>(0, numSymbols), [&](unsigned symbolNumber) {
+        return getAffineDimExpr(symbolNumber + numDims, context);
+      });
+}
+
+/// Convert all symbols in the map to dim expressions, such that the new dim
+/// expressions have positions `numDims, numDims + 1, numDims + 2, ...., numDims
+/// + numSymbols - 1`.
+static AffineMap
+convertDimsToSymbols(AffineMap map, unsigned numDims, unsigned numSymbols,
+                     SmallVector<AffineExpr> &symbolReplacements) {
+  return map.replaceDimsAndSymbols(/*dimReplacements=*/ArrayRef<AffineExpr>{},
+                                   symbolReplacements, numDims + numSymbols, 0);
+}
+static SmallVector<AffineMap>
+convertDimsToSymbols(ArrayRef<AffineMap> maps, unsigned numDims,
+                     unsigned numSymbols,
+                     SmallVector<AffineExpr> &symbolReplacements) {
+  return llvm::map_to_vector(maps, [&](AffineMap map) {
+    return convertDimsToSymbols(map, numDims, numSymbols, symbolReplacements);
+  });
+}
+static SmallVector<AffineMap> convertDimsToSymbols(MLIRContext *context,
+                                                   ArrayRef<AffineMap> maps,
+                                                   unsigned numDims,
+                                                   unsigned numSymbols) {
+  auto symbolReplacements = getDimExprsForSymbols(context, numDims, numSymbols);
+  return convertDimsToSymbols(maps, numDims, numSymbols, symbolReplacements);
+}
+
+SmallVector<Range> CustomOp::getIterationDomainForDimensions(
+    OpBuilder &builder, ArrayRef<unsigned> dims, ArrayRef<unsigned> symbols) {
+  CustomOp customOp = *this;
+  SmallVector<AffineMap> maps = customOp.getIndexingMapsArray();
+  if (maps.empty()) {
+    return SmallVector<Range>{};
+  }
+
+  Location loc = getLoc();
+
+  // 1. Create a flat list of all the operand shapes (similar to Linalg)
+  SmallVector<OpFoldResult> allShapesSizes =
+      createFlatListOfOperandDims(builder, loc, customOp);
+
+  // 2a. Next we need to get a map from shapes to loop. Since `custom_op`
+  //     has indexing maps that have symbols, to make this work correctly
+  //     compute new maps that replaces the symbols with "new" dims.
+  unsigned numDims = getNumLoops();
+  unsigned numSymbols = getNumNonLoopDimensions();
+  MLIRContext *context = getContext();
+  SmallVector<AffineMap> modifiedMaps =
+      convertDimsToSymbols(context, maps, numDims, numSymbols);
+
+  // 2b. Concat the affine maps.
+  AffineMap concatMap = inversePermutation(concatAffineMaps(modifiedMaps));
+  SmallVector<Range> ranges;
```

Comment:
The [inversePermutation](https://github.com/llvm/llvm-project/blob/b3c1403dc00a41c8222d4b1dd82cbb7c56fd61ae/mlir/lib/IR/AffineMap.cpp#L791-L813) could return an empty map. What happens if it fails? Do we verify it in op verifier?